WIP: Issue75 - Diffviewer, Reshuffle some links

Dottori opened the pull request:
Repositioned and resized links in Diff Viewer

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/firefox-code-coverage-frontend/93)
<!-- Reviewable:end -->

gmierz commented on src/components/diffviewer.js:
```diff
@@ -86,6 +86,14 @@ const DiffViewer = ({ appError, coverage, node, parsedDiff, summary }) => (
+      {...coverage.parentMeta(coverage)}
+      {...coverage.diffMeta(node)}
+      coverage={coverage}
+
```
Remove unnecessary blank line.

gmierz commented on src/style.css:
```diff
@@ -44,6 +44,18 @@ pre { margin: 0; }
 	max-width: 600px;
 }
+.meta-footer {
+		text-align: center;
```
Should be aligned to properties in 'meta-footer a{' like 'color'.

gmierz commented on src/components/diffviewer.js:
```diff
@@ -86,6 +86,14 @@ const DiffViewer = ({ appError, coverage, node, parsedDiff, summary }) => (
       ))}
+    {(coverage) &&
+      <DiffFooter
+      {...coverage.parentMeta(coverage)}
```
91-93 should be indented like shown here: https://github.com/mozilla/firefox-code-coverage-frontend/blob/master/src/components/diffviewer.js#L70

gmierz commented on src/components/diffviewer.js:
```diff
@@ -96,23 +104,26 @@ const CoverageMeta = ({ ccovBackend, codecov, coverage, gh, hgRev, pushlog, summ
+    <span>
+      <a href={pushlog} target="_blank">Push Log</a>&nbsp;
+      <a href={gh} target="_blank">GitHub</a>&nbsp;
+      <a href={codecov} target="_blank">codecov.io</a>&nbsp;
```
Should be 'Codecov' not 'codecov.io'. Also 'Push Log' and 'Hg Diff' (please capitalize the 'd' to make the links have similar formatting) should stay at the top.

gmierz replied:
You can ignore the capitalize comment, you've already fixed it. :)

armenzg commented:
At this point I would also remove "Parent meta" and "Changeset meta".
Also drop the "Current coverage" line.

Align it all to the left.

You can see the deployment of the PR here:
https://firefox-code-coverage-pr-93.herokuapp.com/#/changeset/a937ea520febcc4175e49d5d137bfe48d3d607c2

Dottori commented:
Hi Armen, in regards to removing "Parent meta" and "Changeset meta", I was advised to keep push log and hg diff at the top of the page (comments on lines 117-120 in the changes requested by @gmierz). Would I still want to drop the current coverage line and remove the parent meta and changeset meta given this request? 

Also, could you elaborate on what you mean by dropping the current coverage line? 

I have pushed some changes. Please let me know if further changes have to be made in order to remove/reposition the meta data.

armenzg commented:
I see a bunch of changes on this PR that don't have to do with what you're changing.
Would you mind pulling new changes to your master repo and rebase your branch?
```
git checkout master
git pull git@github.com:mozilla/firefox-code-coverage-frontend.git master
git checkout issue75
git rebase master
git push
```
If you visit your fork you can notice this text: 'This branch is 2 commits behind mozilla:master. '

Back to the PR, I was referring to *just* removing the strings "parent meta" and "changeset meta". You can keep the rest of the contents.

Side note, could you please see how it looks if the CSS class `coverage-meta` was to spread with the width of the page? All centered. 

Also, I've noticed that for a moment we see the three bottom links show up at the same time while the coverage code is loading up and it does not look pretty. Perhaps we can even try to prevent showing those 3 links under the coverage data fetch is completed:
![✂-1](https://files.reviewableusercontent.io/images/reviewable/44410/f20CUa907zCKA935bsRckne0.png)

---

Reviewed 5 of 11 files at r2.
Review status: 5 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/style.css, line 48 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L3UP7X6teCXQ_UAQVtn-r1-48:-L41EnafhIjK12e9izzC:b6zrrkh) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/75db7ca9c1e7f45ad20651505680a981607ac588/src/style.css#L48)):*
<details><summary><i>Previously, gmierz wrote…</i></summary><blockquote>

Should be aligned to properties in 'meta-footer a{' like 'color'.
</blockquote></details>

Good catch.

---

*[src/style.css, line 53 at r2](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L41FZshXcku11LrPg8Z:-L41FZsiC90EDmLezyQh:b3yvvha) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/576c6b658e30e060e28edc188287956e5fae9a3e/src/style.css#L53)):*
> ```CSS
> 	color: gray;
> 	border-radius: 10px;
> 	display: table-cell;
> ```

If you use Firefox's inspector you can select an element and enable/disable CSS properties.
`display: 'table-cell` does not seem to be doing anything important.

---

*[src/style.css, line 56 at r2](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L41FmwvZslU9RTkyd9w:-L41FmwvZslU9RTkyd9x:b-9rwt6r) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/576c6b658e30e060e28edc188287956e5fae9a3e/src/style.css#L56)):*
> ```CSS
> 	text-align: center;
> 	padding-left: 10px;
> 	padding-right: 10px;
> ```

Try to write this footer using flexbox. It is a good skill to gain.
Try to accomplish with `justify-content` what you're trying to do with the padding:
https://css-tricks.com/snippets/css/a-guide-to-flexbox/#article-header-id-6

Take your time to learn this technique.

Also, I believe you can write this previous CSS class like this:
```css
.meta-footer {
   text-align: center,
   a {
      ...
   }
}
```
Give it a try and see if I got it right.

---

*[src/components/diffviewer.js, line 122 at r2](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L41Q3A4Ceq-eO_7zT2u:-L41Q3A51ifMr_1J5k_c:b-dven06) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/576c6b658e30e060e28edc188287956e5fae9a3e/src/components/diffviewer.js#L122)):*
> ```JavaScript
> const DiffFooter = ({ gh, codecov, ccovBackend }) => (
>   <div className="meta-footer">
>     <span>
> ```

You don't need this `span`.

---

*[src/components/diffviewer.js, line 123 at r2](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L41Q6UoSiZkmA3LSUbd:-L41Q6UoSiZkmA3LSUbe:b-1768rv) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/576c6b658e30e060e28edc188287956e5fae9a3e/src/components/diffviewer.js#L123)):*
> ```JavaScript
>   <div className="meta-footer">
>     <span>
>       <a href={gh} target="_blank">GitHub</a>&nbsp;
> ```

Once you use flexbox you won't need the `&nbsp;`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-:-L41D-8SoWSMey4DPrNK:b38sbbb)*
<!-- Sent from Reviewable.io -->

Dottori commented:
@armenzg : 

Hi Armen, I have made changes to my pull request. Hopefully I have not missed any of the requested changes.

Flexbox appears to be able to justify content, but all of the options appear to utilize the entire width of the div which appears a little awkward since there are only three elements. In the end, I used flexbox and justified the content, but I also changed the width of .meta-footer. Please let me know if you would prefer having the links evenly dispersed across the full width of the page.  

I was unable to nest the css as recommended above for 'meta-footer' and 'a' as it did not seem to apply the css rules when loading the page. 

I have added code to render the footer only after the DiffViewer finishes parsing the changeset diffs. Since this is my first time utilizing JSX component states to update the DOM, please let me know if my code is inefficient or can be improved.

armenzg commented:
Hi Dottori,
I have another thought for this PR. I've noticed that I keep clicking on the "bugzilla link" rather than on the link that will take me to the diff viewer. This is a usability issue. Because we use two links with the same styling it's hard to tell that there are two different links:
![✂-1](https://files.reviewableusercontent.io/images/reviewable/44410/wN-VBYoT_ZcrECJj02FFXdMD.png)

I think this could be fixed with a small icon at the end of the description (not make the text a link) and make the icon to be the hyperlink. I won't ask for this in this issue but we should file it to tackle in the future.
I even found a perfect icon! https://github.com/fitzgen/bugzilla-todos/blob/gh-pages/lib/bugzilla.png

The top and bottom divs look good.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/components/diffviewer.js, line 89 at r2](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-L46Q_taYUH_uiy33Kq-:-L46Q_tbIUooYKQbTk7k:bumcgvf) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/576c6b658e30e060e28edc188287956e5fae9a3e/src/components/diffviewer.js#L89)):*
> ```JavaScript
>         />
>       ))}
>     {(coverage) &&
> ```

You don't need to add `diffLoaded`. We should actually debug why using `coverage` was not good enough.
If you use the dev tools debugger you should be able to determine why checking `coverage` was allowing to show the div. Let me know if you would need to know how to use the inspector/debugger.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-:-L46RK0VEERd7UY4w86C:b-fylkml)*
<!-- Sent from Reviewable.io -->

Dottori commented:
Hi armenzg,
The icon sounds like a good idea =D. Would you prefer for it to be on the left-hand side of the description? 

I was actually wondering about the React Component Lifecycle. To my understanding, componentDidMount() is called after the render method is executed. csetMeta is initially instantiated to be undefined. so initially the page renders with only the return to main page link. 

The following section confused me because it appears that coverage is updated, which updates the diffviewer component. For some reason at this point the header with the metadata loads, fetchSetCoverageData executes,, and the page renders again. It's only afterwards that parsedDiff is updated and the page is rendered one more time, which is why coverage cannot be used (coverage is assigned a value before parsedDiff finishes parsing). While line 29 (the Promise.all()) require both fetchSetCoverageData and fetchSetDiff to resolve before resolving componentDidMount, it seems like it's still triggering updates to the DOM? Again, I'm still a beginner with JSX and frontend work so please let me know if my terminology/understanding is incorrect.

Dottori commented:
@armenzg Could I instead use:

    {(coverage && parsedDiff.length > 0) &&
      <DiffFooter
        {...coverage.parentMeta(coverage)}
        {...coverage.diffMeta(node)}
        coverage={coverage}
      />}

Dottori commented:
Or actually even just parsedDiff.length > 0 since coverage should not evaluate to false when parsedDiff is populated?

armenzg commented:
@Dottori `parsedDiff.length > 0` should do the trick 👍

Dottori commented:
I have made the requested changes =)~.

armenzg commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :call_me_hand: 

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/93#-:-L4BHrNWDiKsv9Zu-NIr:b89604m)*
<!-- Sent from Reviewable.io -->
